### ISSUE TITLE
--no-optional and package.json

### ISSUE THREAD

**Issue by msmolka** [npm/npm#14185]
As far as I concerned there is currently no way to define in package.json which package should be installed with --no-optional command. When I install package as
npm install --save --no-optional fabric 
then all is fine
but then during install or other developer will grab project the 
npm install will install all packages from package.json with optional dependencies.

How to preserve such option in project

**Comment by legodude17** [comment]
No `--no-optional` installs the package without the optional dependencies. Why do you need to install different packages other than "not optional ones"?

**Comment by msmolka** [comment]
That is the problem. I don't want to. But this value is not preserved in package.json file. So when I run npm install to update all packages, or restore in new environment, the package that should be installed without optional dependencies is installed with them. I want to prevent to install optional dependencies permanently. Preserve somewhere this option.

**Comment by legodude17** [comment]
So you want to be able to always install one of your deps without the optional, even when `--no-optional` is not specified.

**Comment by msmolka** [comment]
Yes, I mean there is no way to specify this parameter inside dependencies tree in packages.json file. 
The scenario:
1. adding new required package to project:
npm install --save --no-optional fabric 
2. Package is saved under dependencies and properly installed and working
3. Other developer grab the project and trying restore missing dependencies
4. Runs npm install
5. Because there is no info that package does not require dependencies to install, it fails to install (e.g. in this example fabric requires Cairo. But in my project I just require plain js, no additional builds so installation without dependencies is the best. 
6. Developer has broken environment, and needs to manually uninstall fabric and install it manually using --no-optional command.

**Comment by legodude17** [comment]
Technically, when Cairo fails to install, since it is an optional dep, then
npm should just continue without it.

**Comment by msmolka** [comment]
It continues, but on every project build there is a lot of errors and strange external dependencies

**Comment by legodude17** [comment]
Can you post a basic github repo that demos that?

**Comment by msmolka** [comment]
[TestPackage.zip](https://github.com/npm/npm/files/518293/TestPackage.zip)
This folder has one package installed with following command 
npm install --save --no-optional fabric 

It had only one folder inside node_modules folder called fabric
Now when you go to project folder
and run "npm update" or "npm install" just to restore this single one folder
it will install a lot of dependencies. and fail because of global system related stuff missing.
It will contain folder, but it fact whole folder will be messy
I realize this is semi related to Node.js itself. I'm just feeling that preserving this flag somehow for saved dependencies would very useful.

**Comment by scholtzm** [comment]
Very similar situation with this package: [qrious](https://github.com/neocotic/qrious#nodejs-dependencies)

This package requires`node-canvas` when running under node.js but has no external dependencies when running in browser.

Currently there is no way to specify this flag in package.json for this specific dependency and thus every install leads to trying to install `node-canvas` even though it's never used.

**Comment by Cyken-Zeraux** [comment]
Also have the problem where node-canvas is under optional dependencies for Fabric library, which does not compile without python and a few other things that aren't used in the browser. Support for attaching options to dependencies would solve a bunch of edge cases like this.

**Comment by ramijarrar** [comment]
We are also facing the same problem when [trying to use](https://github.com/BrowserSync/browser-sync/issues/983) the 'browser-sync' package.

**Comment by mariotacke** [comment]
Same stuff, my team is complaining that `fabric`s optional dependencies fail to install. IMO, it is reasonable to expect devs to run `npm install` but not `npm install --no-optional` every time you install a new package.

**Comment by hendrul** [comment]
In my case a utility package uses optionally nodegit, and it's not only unused most of the time, but also takes a lot of time to install because it uses node-gyp

**Comment by born2net** [comment]
same issue here we need a way to specifically set no options for fabric not for the entire and npm install.. still no solution?

**Comment by mariotacke** [comment]
@born2net We should probably ask fabric.js (@kangax) to mark canvas/cairo as a peer dependency. Not sure if that would bode well for existing CI. Being able to mark a package to be installed without optionals would be nice to have though.

**Comment by born2net** [comment]
yes I asked fabric guys to remove optionalDeps and just have readme for node-canvas instead.
hopefully npm will support --no-optional per package.json

**Comment by diracdeltas** [comment]
+1, we are running into this issue with `npm install --save --no-optional qrious`

**Comment by dobriai** [comment]
Yes! We have the same problem. Module `grunt-contrib-compress` v1.4.1 has some optional dep on `iltorb`, which requires some compilation with python < 3.5. Installation tries to build, fails and continues OK, but wastes time for something we know we don't need. It also pollutes the log, which is the worst part! Thus, I would like to be able to specify in some config file that for this particular module I want 'no-dependencies' installation.

**Comment by hendrul** [comment]
@dobriai 

> I would like to be able to specify in some config file that for this particular module I want 'no-dependencies' installation.

I think it can only be specified in package.json right next to the dependency if we want to keep our dependencies in one place, so we don't have to read different places or files to understand the dependency graph of our app

**Comment by born2net** [comment]
I wish there was away to remove the optional dependencies without forking the project...

**Comment by hendrul** [comment]
Something like:
```json
{
  "dependencies": {
    "dependencyA": {
      "version": "^0.9.0",
      "optionals": false
    },
    "dependencyB": {
      "url": "git+http://host/user/dependency-b.git",
      "optionals": false
    }
  }
}
```

**Comment by hendrul** [comment]
This optional flag is not meant to be recursive, should only apply to that module only

**Comment by mariotacke** [comment]
@hendrul do you think we can design it to be backwards compatible? Not sure if introducing an object instead of a string would break old clients. Maybe we can get away with a modifier such as `~` or `^`. Really just spit-balling here.

Example: 
```
{
  "dependencies": {
    "dependencyA": "@^0.9.0"
  }
}
```

or

```
{
  "dependencies": {
    "dependencyA": "^0.9.0?no-opt"
  }
}
```

**Comment by hendrul** [comment]
@mariotacke what would happend with urls, or local paths? urls already have meaning for querystrings

**Comment by mariotacke** [comment]
You are right. I guess something like this would be a breaking change. Maybe a completely different key in the package.json?

```
{
  "dependencies": {
    "dependencyA": "^0.9.0"
  },
  "options": {
    "dependencyA": {
      "optionals": false,
      ... other options ...
    }
  }
}
```

**Comment by kirrg001** [comment]
+1

**Comment by cihadhoruzoglu** [comment]
+1

**Comment by legodude17** [comment]
@kirrg001 @cihadhoruzoglu please don't add `+1` comments, instead you can vote on the issue comments.

**Comment by owen26** [comment]
And here is what happened when we don't add +1 to the comments...

**Comment by scholtzm** [comment]
@zkat Is this issue supposed to be closed?

**Comment by vshymanskyy** [comment]
Oh wow. Why is this still closed?

**Comment by Cuel** [comment]
This is badly needed with angular-cli and node-sass since it fetches the binary and dies due to corporate firewalls

```
npm install @angular/cli@1.3.1 --no-optional
```

works fine, node-sass is not added to package-lock
next `npm install` adds it...

**Comment by andmis** [comment]
It is a bit ridiculous that NPM/`package.json` doesn't have this functionality.

**Comment by oracle000** [comment]
What happen to this issue ? Is there's a solution for this already ?

**Comment by ljharb** [comment]
@KenanY can this be reopened?

**Comment by zkat** [comment]
Reopening and re-triaging as a feature request

**Comment by nathanredblur** [comment]
It's no so hard to do, when going to have this done ?

**Comment by palashkulsh** [comment]
Facing same problem when including https://www.npmjs.com/package/kafka-node package in my package.json

**Comment by bluelovers** [comment]
same here

```
npm WARN epub@0.2.0 requires a peer of zipfile@^0.5.0 but none is installed. You must install peer dependencies yourself.
```

i wanna 

```
"peerDependencies": {},
  "optionalDependencies": {
    "zipfile": "^0.5.0"
  },
```

not

```
"optionalDependencies": {},
  "peerDependencies": {
    "zipfile": "^0.5.0"
  },
```

and don't try install zipfile when install epub without --no-optional

**Comment by matteocontrini** [comment]
Any chance that this can be actually considered? It's been (at least) a year and a half with no indication on whether this is going to be implemented or not at some time

**Comment by maxcbc** [comment]
Yes please consider this. At bare minimum can we have something in npm shrinkwrap which says 'skip this dependency'. I've got a project with a vulnerability in an optional dependency and I have no way of removing it, except via a post install 'rm -rf node_modules/<dependency>'

**Comment by kopax** [comment]
I want to create a lib that export a different component if my user is using **react-intl**

I can't do it because the **optionalDependency** is always installed on the user project when people install my package.

**Comment by kopax** [comment]
@npm-robot can you please ask npm humans to give us an answer for this x?)

**Comment by legodude17** [comment]
I would suggest you post something about this in https://npm.community/c/ideas.
